File: 3699. Number of ZigZag Arrays I.cpp
class Solution {
public:
    static const int MOD = 1000000007;

    int zigZagArrays(int n, int l, int r) {
        int m = r - l + 1;

        vector<long long> dp(m, 1);

        for (int len = 2; len <= n; len++) {
            reverse(dp.begin(), dp.end());

            long long prefixSum = 0;

            for (int i = 0; i < m; i++) {
                long long current = dp[i];

                dp[i] = prefixSum;

                prefixSum = (prefixSum + current) % MOD;
            }
        }

        long long totalWays = 0;

        for (long long ways : dp) {
            totalWays = (totalWays + ways) % MOD;
        }

        return (int)((totalWays * 2) % MOD);
    }
};

File: 3699. Number of ZigZag Arrays I.js
const MOD = 1000000007;

/**
 * @param {number} n
 * @param {number} l
 * @param {number} r
 * @return {number}
 */
var zigZagArrays = function(n, l, r) {
    const m = r - l + 1;

    let dp = new Array(m).fill(1);

    for (let len = 2; len <= n; len++) {
        dp.reverse();

        let prefixSum = 0;

        for (let i = 0; i < m; i++) {
            const current = dp[i];

            dp[i] = prefixSum;

            prefixSum = (prefixSum + current) % MOD;
        }
    }

    let totalWays = 0;

    for (const ways of dp) {
        totalWays = (totalWays + ways) % MOD;
    }

    return (totalWays * 2) % MOD;
};
